### ISSUE TITLE
Forbid private commit in tsp-location.yaml

### ISSUE THREAD

**Issue by pshao25** [Azure/azure-sdk-for-net#42860]
We currently allow user to pass in the commit from an arbitrary branch (not from main) from azure-rest-api-specs. Then when we update the spec in main branch, the change in that arbitrary commit lost, which causes breaking change. We should forbid user pass in any commit not from main.

**Comment by jsquire** [comment]
I'm not sure that I agree that we can/should block this as a general rule.   Doing so would block local testing or testing via feature branch until a spec was finalized, which are valid scenarios.  While I do agree that it would be potentially good to block merging to our `main` with a location pointing to another branch in the specs repo, but I'm also not sure there aren't valid cases for that.

This is something that can/should be caught by code reviews.  If our team is generating or reviewing, having it be part of our review should be sufficient.  If a service team has full responsibility for their library, this would be something they would need to evaluate as part of their process.

Can you help me understand what scenario prompted you to open this issue and what you think should be blocking?  I'm not sure if you're suggesting a build verification step or something different.


//cc: @m-nash for additional thoughts.

**Comment by pshao25** [comment]
`Azure.AI.DocumentIntelligence` is an example. The current commit is https://github.com/Azure/azure-sdk-for-net/blob/6c20209b70e8458bbba848f9b323cad7a58613e9/sdk/documentintelligence/Azure.AI.DocumentIntelligence/tsp-location.yaml#L2, which comes from an un-merged [PR](https://github.com/Azure/azure-rest-api-specs/pull/28032).

We currently need to do some change in their spec, which is expected to be into main and it is already in main. Then when we update the commit to the latest, the change in that un-merged PR lost.

> Doing so would block local testing or testing via feature branch until a spec was finalized, which are valid scenarios.

Yeah, I think we should keep local testing as well. Maybe we could make "checking whether it comes from main" an optional CI rule, so that when we do the review we could easily figure out. An if we want an exception for special case, it won't block users.

**Comment by kinelski** [comment]
I agree that making this optional would be ideal.

To give you some context on Document Intelligence, the service usually merges spec changes just a couple of days before the SDK release, so I kept .NET-specific changes to the spec in a separate branch for testing and intended to merge it after the service's PR came in.

When opening a PR, however, while the generation targeting my branch worked fine in my local environment, I got failures on the TypeSpec Validation pipeline that prevented me from merging it (I still don't know why). This happened 2 week days before the release date.

I can see that issue happening again given the short time span we usually have between the service specs getting merged and the SDK release date. Forcing generation from the main branch could cause delays in future releases.

**Comment by jsquire** [comment]
@pshao25 : Aside from development scenarios, the big problem with your proposal is that it also breaks support paths like hotfixes, where we will often intentionally pull a tag, make a specific change, and release just that change.   

Technically, this is also a challenge as git history is non-linear and does not explicitly track the branch for a commit.   The closest that I believe you'd be able to come would be to find which branches have that commit anywhere in their history and assume that if `main` is in that list, things are fine.  In my opinion, there is not enough value offered by doing this to justify the work that would be needed.

I'll leave this open for another week to collect additional perspectives, but unless this is something that you'd like to take on, I do not see us allocating resources to do this work.

**Comment by pallavit** [comment]
> I'll leave this open for another week to collect additional perspectives, but unless this is something that you'd like to take on, I do not see us allocating resources to do this work.

+1 to @jsquire thoughts, I think there is a valid use case here, so this should not be fully blocked.

**Comment by pshao25** [comment]
@jsquire @pallavit My proposal is just on `main` so it won't block any hotfixes which should be from some release branches. It could be optional too.

I heard @chunyu3 and @raych1 are already working on this, so they will take on this.

**Comment by jsquire** [comment]
@pshao25, @chunyu3, @raych1:  Please ensure that any PR for these changes is marked for review by all members of the Azure SDK for .NET team and does not merge without explicit approval from Pallavi or myself.

**Comment by raych1** [comment]
@jsquire , I implemented commit validation to ensure that commits are from the main branch of the Azure spec repo in the `release` pipeline. However, this validation hasn't been added to the CI/PR pipeline. I remember that the reason is to enable the feature development which might use the spec from other branches.

**Comment by jsquire** [comment]
@raych1:  Is there a PR that we can review?   Please do not add anything to any pipeline without the .NET team being able to review and sign-off.   These changes should not be made in a vacuum as they impact the entire .NET SDK ecosystem.

**Comment by jsquire** [comment]
I found the related PR, which should not have been merged as there was no approval from a .NET team member, only EngSys.    
https://github.com/Azure/azure-sdk-for-net/pull/40783

I'm going to close this issue out, as there hasn't been any noted difficulties with release since the PR was merged.   Going forward, please be aware that these changes do require architect/manager approval.

**Comment by pshao25** [comment]
@jsquire That PR was merged because I had no choice. I fixed the problem in that PR by trying to understand his intention, which I think is not the correct process. And I did the same thing to another PR having the same problem. I opened up this issue to prevent such thing from happening again. Do you have a better solution? 

+ @lirenhe

**Comment by lirenhe** [comment]
Just share my thinking on this:
1. The majority of services today should use PR links from public spec repo, most of the links to private spec repo we found were mistakenly merged.
2. Leveraging tooling/pipeline support could help to increase efficiency of the reviewers by reducing the manual effort.
3. there could be exceptions, we could have a suppression process without compromising the overall flow. We already did that in PRs for spec repo.
